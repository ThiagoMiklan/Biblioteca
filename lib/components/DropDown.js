"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var React = _interopRequireWildcard(require("react"));

var _Button = _interopRequireDefault(require("./Button"));

var _Icon = _interopRequireDefault(require("./Icon"));

var _type_validations = require("../tools/type_validations.js");

var props_obj = {
  definition: 'string',
  itens: 'array',
  onClick: 'function'
};

var DropDown = function DropDown(props) {
  (0, _type_validations.validate)(props, props_obj, "DropDown");
  var definition = props.definition == undefined ? "dropdown" : "dropdown " + props.definition;
  return /*#__PURE__*/React.createElement("div", {
    className: definition
  }, /*#__PURE__*/React.createElement("div", {
    "class": "dropdown-trigger"
  }, extractButton(props.itens, props.onClick)), /*#__PURE__*/React.createElement("div", {
    "class": "dropdown-menu"
  }, /*#__PURE__*/React.createElement("div", {
    "class": "dropdown-content"
  }, assemble(props.itens))));
};

function extractButton(itens, onClick) {
  var name = "";
  var icon_name = "";
  var icon_definition = "";

  if (itens != undefined && Array.isArray(itens)) {
    if (itens[0] != undefined) {
      var item = itens[0];
      name = item["value"];
      icon_name = item["icon_name"];
    }
  }

  var label = "";

  if (icon_name != undefined && icon_name != '') {
    label = /*#__PURE__*/React.createElement(_Icon["default"], {
      icon_name: icon_name,
      right: true
    }, name);
  } else {
    label = /*#__PURE__*/React.createElement(_Icon["default"], {
      icon_name: "fas fa-angle-down",
      definition: icon_definition,
      right: true
    }, name);
  }

  return /*#__PURE__*/React.createElement(_Button["default"], {
    onClick: onClick,
    label: label
  });
}

function assemble(itens) {
  var code = /*#__PURE__*/React.createElement(React.Fragment, null);

  if (itens != undefined && Array.isArray(itens)) {
    code = itens.map(function (item) {
      return decideTypeOfItens(item);
    });
  }

  return code;
}

function decideTypeOfItens(item) {
  var item_code = "";

  if (item["link"] == true) {
    item_code = assembleItensLinks(item);
  } else {
    item_code = assembleComponentsItens(item);
  }

  return item_code;
}

function assembleItensLinks(item) {
  var definition = item["definition"] == undefined ? "dropdown-item" : "dropdown-item " + item["definition"];
  return /*#__PURE__*/React.createElement("a", (0, _extends2["default"])({
    className: definition,
    onClick: item["onClick"],
    href: item["href"]
  }, item["custom"]), item["value"]);
}

function assembleComponentsItens(item) {
  var definition = item["definition"] == undefined ? "dropdown-item" : "dropdown-item " + item["definition"];
  return /*#__PURE__*/React.createElement("div", (0, _extends2["default"])({
    className: definition,
    onClick: item["onClick"],
    href: item["href"]
  }, item["custom"]), item["value"]);
}

var _default = DropDown;
exports["default"] = _default;