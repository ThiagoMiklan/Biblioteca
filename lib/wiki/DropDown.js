"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _react = _interopRequireWildcard(require("react"));

var _reactDom = _interopRequireDefault(require("react-dom"));

var _DropDown = _interopRequireDefault(require("./components/DropDown"));

require("bulma/css/bulma.css");

var itens_link = [{
  value: "Dropdown Button",
  link: true,
  href: "https://ufsc.br",
  onClick: function onClick() {
    alert("Clicou");
  }
}, {
  value: "Dropdown Item",
  link: true,
  href: "https://ufsc.br"
}, {
  value: "Other Dropdown Item",
  link: true,
  href: "https://ufsc.br"
}, {
  value: "Active Dropdown Item",
  link: true,
  definition: "is-active",
  href: "https://ufsc.br"
}, {
  value: "Other Dropdown Item",
  link: true,
  href: "https://ufsc.br"
}, {
  value: "With a divider",
  link: true,
  href: "https://ufsc.br"
}];
var itens_div = [{
  value: /*#__PURE__*/_react.default.createElement("p", null, "You can insert ", /*#__PURE__*/_react.default.createElement("strong", null, "any type of content"), " within the dropdown menu."),
  onClick: function onClick() {
    alert("Clicou");
  }
}, {
  value: /*#__PURE__*/_react.default.createElement("p", null, "You simply need to use a ", /*#__PURE__*/_react.default.createElement("code", null, "<div>"), " instead.")
}, {
  value: "This is a link",
  link: true
}];

var HookDropDown = function HookDropDown(props) {
  var initial = {
    definition: ""
  };

  var _useState = (0, _react.useState)(initial),
      _useState2 = (0, _slicedToArray2.default)(_useState, 2),
      state = _useState2[0],
      setState = _useState2[1];

  var onClick = function onClick() {
    if (state.definition == "is-active") {
      setState({
        definition: ""
      });
    } else {
      setState({
        definition: "is-active"
      });
    }
  };

  return /*#__PURE__*/_react.default.createElement(_DropDown.default, {
    definition: state.definition,
    onClick: onClick,
    itens: itens_link
  });
};

_reactDom.default.render( /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_DropDown.default, {
  definition: "is-active",
  itens: itens_link
}), /*#__PURE__*/_react.default.createElement(_DropDown.default, {
  definition: "is-hoverable",
  itens: itens_link
}), /*#__PURE__*/_react.default.createElement(_DropDown.default, {
  definition: "is-up is-active",
  itens: itens_link
})), document.getElementById('root'));