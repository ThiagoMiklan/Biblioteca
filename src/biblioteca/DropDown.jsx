// @flow
import * as React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import getClassName from '../tools/getClassName';
import Button from '../biblioteca/Button';
import Icon from '../biblioteca/Icon';
import checkerDefinition from '../tools/checker.js';



/**
 * 
 * Observações: 
 * - O primeiro item do DropDown é gerado através do primeiro item de itens(props.itens)
 * - Para que os itens sejam exibidos ao clicar no DropDown é necessário tratar a características
 *  active (que define is-active do Bulma) através do JavaScript, ao contrário do Hoverable que funciona
 * somente com o CSS
 * 
 * 
 */

 type Props = {
  definition?: string,
  itens?: React.Node,
  onClick?: ()=> void
 }

const DropDown = (props: Props) => {
    /*
      Sério problema ao exibir um dropdown, caso o nome do primeiro não seja um nome muito extenso
      corre o risco dos itens do dropdown não serem exibidos, necessário checagem completa do problema
      para evitar bugs no componente
      !! Problema identificado ao utilizar right(is-right), no caso de erro usado is-active e is-right
      
      Não existe
      */
    checkerDefinition(props.definition,"DropDown");
    var definition = "dropdown " + classnames(getClassName(props.definition, "DropDown"));

    return (<div className={definition}>
                <div class="dropdown-trigger">
                     {extractButton(props.itens,props.onClick)}
                </div>
                <div class="dropdown-menu" id="dropdown-menu" role="menu">
                     <div class="dropdown-content">
                        {assemble(props.itens)}
                     </div>
                 </div>
             </div>
             );
}

// Monta um botão a partir do primeiro item da lista
function extractButton(itens,onClick) {
    var name = "";
    var icon_name = ""
    var icon_definition = "";

    // extrai o primeiro item da lista que foi informada como props
    if (itens != undefined && Array.isArray(itens)) {
        if(itens[0] != undefined){
            var item = itens[0];
            name = item["value"];
            icon_name = item["icon_name"];
        }
     }
    /**
     * Funcionamento da renderização condicional
     * Se o nome do icone não é indefinido ou vazio, é usado o icone default para o dropdown
     * Senão, signigica que foi informado um nome para o icone, então é usado o que foi passado
     */
    return <Button onClick={onClick} >
        {(icon_name != undefined && icon_name != '')
            ? <Icon icon_name={classnames(getClassName(icon_name, "DropDown"))} icon_right={true}>{name}</Icon>
            
            : <Icon icon_name={classnames(getClassName("default-icon", "DropDown"))} definition={icon_definition} icon_right={true} >{name}</Icon>
        }
    </Button>
}

function assemble(itens) {
    var code = <></>
    if (itens != undefined && Array.isArray(itens)) {
        code = itens.map(item => decideTypeOfItens(item));
    }

    return code;
}

function decideTypeOfItens(item) {
    var item_code = "";
    if (item["link"] == true) {
        item_code = assembleItensLinks(item)
    } else {
        item_code = assembleComponentsItens(item)
    }

    return item_code;
}

function assembleItensLinks(item) {
    var definition = classnames(getClassName("dropdown-item " + item["definition"], "DropDown"));
    return <a className={definition} onClick={item["onClick"]} href={item["href"]}>{item["value"]}</a>;
}

function assembleComponentsItens(item) {
    var definition = classnames(getClassName("dropdown-item " + item["definition"], "DropDown"));
    return <div className={definition} onClick={item["onClick"]} href={item["href"]} >{item["value"]}</div>;
}

DropDown.propTypes = {
    definition: PropTypes.string,
    itens: PropTypes.array,
    onClick: PropTypes.func
}

export default DropDown;