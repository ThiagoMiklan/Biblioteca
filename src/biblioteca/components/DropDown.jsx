import React from 'react';
import PropTypes from 'prop-types';
import classname from 'classnames';
import getClassName from '../../tools/getClassName';
import Button from '../elements/Button.jsx';
import { renderToString } from 'react-dom/server';

class DropDown extends React.Component {

    render() {
      /*
      Sério problema ao exibir um dropdown, caso o nome do primeiro não seja um nome muito extenso
      corre o risco dos itens do dropdown não serem exibidos, necessário checagem completa do problema
      para evitar bugs no componente
      !! Problema identificado ao utilizar right(is-right), no caso de erro usado is-active e is-right
      
      Não existe
      */
        var definition = "dropdown " + classname(getClassName(this.props.definition, "DropDown"));
        var dropdown_name = this.props.dropdown_name;
        var code = "";
        return (
          <div class={definition}>
                <div class="dropdown-trigger">
                    {this.props.button_dropdown}
                </div>
                <div class="dropdown-menu" id="dropdown-menu" role="menu">
                    <div class="dropdown-content">
                        {assemble(this.props.itens)}
                    </div>
                </div>
             </div>
             

        );

       
    }

}

function assemble(itens) {
  return itens.map(item => decideTypeOfItens(item));
}

function decideTypeOfItens(item) {
    // link é o default
    var item_code = "";
    if (item["link"] == true) {
        item_code = assembleItensLinks(item)
    } else {
        item_code =  assembleComponentsItens(item)
    }
    var code = renderToString(item_code);
    return item_code;
}

function assembleItensLinks(item) {
    var definition = classname(getClassName("dropdown-item "+item["definition"], "DropDown"));
    return <a className={definition} href={item["href"]}>{item["value"]}</a>;
}

function assembleComponentsItens(item) {
    var definition = classname(getClassName("dropdown-item "+item["definition"],"DropDown"));
    return <div className="dropdown-item">{item["value"]}</div>;
}

DropDown.propTypes = {
    definition: PropTypes.string,
    dropdown_name: PropTypes.string,
    itens: PropTypes.array,
    button_dropdown: PropTypes.element
}

export default DropDown;