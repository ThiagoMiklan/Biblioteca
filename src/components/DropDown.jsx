// @flow
import * as React from 'react';
import PropTypes from 'prop-types';
import Button from './Button';
import Icon from './Icon';
import { validate } from '../tools/type_validations.js';

type Props = {
    definition?: string,
    itens?: React.Node,
    onClick?: () => void
}
const props_obj = {
    definition: 'string',
    itens: 'array',
    onClick: 'function'
}
const DropDown = (props: Props) => {
    validate(props, props_obj);
    var definition = (props.definition == undefined) ? "dropdown" : "dropdown " + props.definition;
    return (<div className={definition}>
        <div class="dropdown-trigger">
            {extractButton(props.itens, props.onClick)}
        </div>
        <div class="dropdown-menu" id="dropdown-menu" role="menu">
            <div class="dropdown-content">
                {assemble(props.itens)}
            </div>
        </div>
    </div>
    );
}

// Monta um botão a partir do primeiro item da lista
function extractButton(itens, onClick) {
    var name = "";
    var icon_name = ""
    var icon_definition = "";

    // extrai o primeiro item da lista que foi informada como props
    if (itens != undefined && Array.isArray(itens)) {
        if (itens[0] != undefined) {
            var item = itens[0];
            name = item["value"];
            icon_name = item["icon_name"];
        }
    }
    /**
     * Funcionamento da renderização condicional
     * Se o nome do icone não é indefinido ou vazio, é usado o icone default para o dropdown
     * Senão, signigica que foi informado um nome para o icone, então é usado o que foi passado
     */
    return <Button onClick={onClick} >
        {(icon_name != undefined && icon_name != '')
            ? <Icon icon_name={icon_name} icon_right={true}>{name}</Icon>

            : <Icon icon_name={"fas fa-angle-down"} definition={icon_definition} icon_right={true} >{name}</Icon>
        }
    </Button>
}

function assemble(itens) {
    var code = <></>
    if (itens != undefined && Array.isArray(itens)) {
        code = itens.map(item => decideTypeOfItens(item));
    }

    return code;
}

function decideTypeOfItens(item) {
    var item_code = "";
    if (item["link"] == true) {
        item_code = assembleItensLinks(item)
    } else {
        item_code = assembleComponentsItens(item)
    }

    return item_code;
}

function assembleItensLinks(item) {
    var definition = (item["definition"] == undefined) ? "dropdown-item" : "dropdown-item " + item["definition"];
    return <a className={definition} onClick={item["onClick"]} href={item["href"]}>{item["value"]}</a>;
}

function assembleComponentsItens(item) {
    var definition = (item["definition"] == undefined) ? "dropdown-item" : "dropdown-item " + item["definition"];
    return <div className={definition} onClick={item["onClick"]} href={item["href"]} >{item["value"]}</div>;
}

DropDown.propTypes = {
    definition: PropTypes.string,
    itens: PropTypes.array,
    onClick: PropTypes.func
}

export default DropDown;