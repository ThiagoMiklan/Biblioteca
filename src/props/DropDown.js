const drop_down_props= {
    "active": "is-active",
    "hoverable": "is-hoverable",
    "right": "is-right",
    "divider": "is-divider"
}

export default drop_down_props;