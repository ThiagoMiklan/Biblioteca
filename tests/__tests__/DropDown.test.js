import React from 'react';
import DropDown from '../../src/components/DropDown';
import renderer from 'react-test-renderer';

var itens_link = [
    { value: "Dropdown Button", link: true, href: "https://ufsc.br", onClick: () => { alert("Clicou") } },
    { value: "Dropdown Item", link: true, href: "https://ufsc.br" },
    { value: "Other Dropdown Item", link: true, href: "https://ufsc.br" },
    { value: "Active Dropdown Item", link: true, definition: "is-active", href: "https://ufsc.br" },
    { value: "Other Dropdown Item", link: true, href: "https://ufsc.br" },
    { value: "With a divider", link: true, href: "https://ufsc.br" }
];

var itens_div = [
    { value: <p>You can insert <strong>any type of content</strong> within the dropdown menu.</p>, onClick: () => { alert("Clicou") } },
    { value: <p>You simply need to use a <code>&lt;div&gt;</code> instead.</p> },
    { value: "This is a link", link: true }
];

test('dropdown normal',()=>{
    const normal = renderer.create(
        <DropDown definition="is-active" itens={itens_link} />
    ).toJSON();

    expect(normal).toMatchSnapshot();
})

test('dropdown with any type of content',()=>{
    const anyContent = renderer.create(
        <DropDown definition="is-active is-right" itens={itens_div} />
    ).toJSON();

    expect(anyContent).toMatchSnapshot();
})


        
        
