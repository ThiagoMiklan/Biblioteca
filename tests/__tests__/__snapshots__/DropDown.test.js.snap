// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dropdown normal 1`] = `
<div
  className="dropdown is-active"
>
  <div
    class="dropdown-trigger"
  >
    <button
      className="button"
      disabled={false}
    >
      <span>
        Dropdown Button
      </span>
      <span
        className="icon "
      >
        <i
          className="fas fa-angle-down"
        />
      </span>
    </button>
  </div>
  <div
    class="dropdown-menu"
  >
    <div
      class="dropdown-content"
    >
      <a
        className="dropdown-item"
        href="https://ufsc.br"
        onClick={[Function]}
      >
        Dropdown Button
      </a>
      <a
        className="dropdown-item"
        href="https://ufsc.br"
      >
        Dropdown Item
      </a>
      <a
        className="dropdown-item"
        href="https://ufsc.br"
      >
        Other Dropdown Item
      </a>
      <a
        className="dropdown-item is-active"
        href="https://ufsc.br"
      >
        Active Dropdown Item
      </a>
      <a
        className="dropdown-item"
        href="https://ufsc.br"
      >
        Other Dropdown Item
      </a>
      <a
        className="dropdown-item"
        href="https://ufsc.br"
      >
        With a divider
      </a>
    </div>
  </div>
</div>
`;

exports[`dropdown with any type of content 1`] = `
<div
  className="dropdown is-active is-right"
>
  <div
    class="dropdown-trigger"
  >
    <button
      className="button"
      disabled={false}
    >
      <span>
        <p>
          You can insert 
          <strong>
            any type of content
          </strong>
           within the dropdown menu.
        </p>
      </span>
      <span
        className="icon "
      >
        <i
          className="fas fa-angle-down"
        />
      </span>
    </button>
  </div>
  <div
    class="dropdown-menu"
  >
    <div
      class="dropdown-content"
    >
      <div
        className="dropdown-item"
        onClick={[Function]}
      >
        <p>
          You can insert 
          <strong>
            any type of content
          </strong>
           within the dropdown menu.
        </p>
      </div>
      <div
        className="dropdown-item"
      >
        <p>
          You simply need to use a 
          <code>
            &lt;div&gt;
          </code>
           instead.
        </p>
      </div>
      <a
        className="dropdown-item"
      >
        This is a link
      </a>
    </div>
  </div>
</div>
`;
